Readme.md
# JQuery Datatables Editor - Alternative to the official one

Add capabilities to add, edit and delete rows in your datatables through the use of modals.

Inline editing is not supported (so far).

## Basic usage

Include in the project the libraries:

1. jQuery
2. [Bootstrap](https://getbootstrap.com/) > 4.0 **or** [Foundation](https://get.foundation/) > 6.0
3. [jquery.dataTables](https://datatables.net/), [dataTables.buttons](https://datatables.net/extensions/buttons/), [dataTables.select](https://datatables.net/extensions/select/)

Define a DataTable as usual.

Pass to the DataTable constructor at least the following arguments: `dom`, `select`, `buttons`, `altEditor: true` (see the examples).

_Bootstrap or Foundation is necessary for action buttons and modals._

## Examples

There are some examples in the folder `example`, for different use cases:

- a DataTable populated and maintained via Javascript, (no AJAX, no databases), with data organized in rows;

- a DataTable populated and maintained via Javascript, (no AJAX, no databases), with data organized in objects;

- a DataTable populated via AJAX, to be used in connection with a databases, with data organized in objects.

Examples can be browsed at <https://luca-vercelli.github.io/DataTable-AltEditor> .

## AJAX Setup

The datatable accepts the following callback functions as arguments:

```
onAddRow(alteditor, rowdata, success, error)
onEditRow(alteditor, rowdata, success, error, originalrowdata)
onDeleteRow(alteditor, rowdata, success, error)
```

In the most common case, these function should call `$.ajax` as expected by the webservice. The two functions `success` and `error` should be passed as arguments to `$.ajax`.

Within the procedures `onAddRow`, `onEditRow`, `onDeleteRow`, you can access datatable object using `alteditor.s.dt`.

Webservice **must** return the modified row in JSON format, because the `success()` function expects this. Otherwise you have to write your own `success()` callback (e.g. refreshing the whole table).

## Row Key

There is no default key in the table. Inside your callback functions, probably you will need a row key to build URL's, in that case you can get them from the `rowdata` parameter.

## Column ID

Please always keep in mind that DataTable framework allows two different kinds of "rows": Arrays and Objects. In first case columns are indexed through integers; in second case columns are indexed by their attribute name. Usually JSON's use the Object approach, but we cannot be sure.

## Column Modifiers, and Validation

Following column options are supported.

Column option                                   | Accepted values                                    | Description
----------------------------------------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------
`type`                                          | `"text" \| "select" \| "hidden" \| ...`            | Type of HTML input to be shown. The value `readonly` is accepted for backward compatibility, but deprecated.
`readonly`                                      | `true \| false`                                    | Add `readonly` HTML attribute
`disabled`                                      | `true \| false`                                    | Add `disabled` HTML attribute
`required`                                      | `true \| false`                                    | Add `required` HTML attribute
`hoverMsg`                                      | `"some msg"`                                       | The message will appear as a tooltip over the input field.
`unique`                                        | `true \| false`                                    | Ensure that no two rows have the same value. The check is performed client side, not server side. Set HTML `"data-unique"` attribute. (Probably there's some issue with this).
`uniqueMsg`                                     | `"some msg"`                                       | An error message that is displayed when the unique constraint is not respected. Set HTML `"data-uniqueMsg"` attribute.
`special`                                       | `"any string"`                                     | Set HTML `"data-special"` attribute (don't know what's that needed for).
`style`                                         | `"any string"`                                     | Set HTML `"style"` attribute.
`editorOnChange`                                | function                                           | Custom onchange function. It will take as arguments the jquery event and the altEditor object.
`value`											| `"some value"`									 | Set default value for add modal. Actual value used for edit modal
**Options for columns with type `"text"`:**     |                                                    |
`pattern`                                       | `r.e.`                                             | The typed text will be matched against given regular expression, before submit.
`msg`                                           | `"some msg"`                                       | An error message that is displayed in case pattern is not matched. Set HTML `"data-errorMsg"` attribute.
`maxLength`                                     | `integer`                                          | Set HTML `"maxlength"` attribute.
`datepicker`                                    | `{}`                                               | Enable a datepicker component. jQuery-UI plugin must be linked. More datepicker configuration options may be passed within the object.
`datetimepicker`                                | `{}`                                               | Enable a datetimepicker component. jQuery datetimepicker plugin must be linked. More datetimepicker configuration options may be passed within the object.
**Options for columns with type `"select"`:**   |                                                    |
`options`                                       | `["a", "b", "c"]` or `{"a":"A", "b":"B", "c":"C"}` | The options that shall be presented.
`select2`                                       | `{}`                                               | Enable a select2 component. Select2 jQuery plugin must be linked. More select2 configuration options may be passed within the object.
`multiple`                                      | `true \| false`                                    | Set HTML `"multiple"` attribute.
`optionsSortByLabel`                            | `true \| false`                                    | Allows select to be sorted by label field.
**Options for columns with type `"textarea"`:** |                                                    |
`rows`                                          | `integer`                                          | Set HTML `"rows"` attribute.
`cols`                                          | `integer`                                          | Set HTML `"cols"` attribute.
**Options for columns with type `"date"`:**     |                                                    |
`dateFormat`                                    | `"YYYY-MM-DD"`                                     | Set date format. Require moment.js library linked.
**Options for columns with type `"number"`**		|													 |	
`step`											| `float`											 | Set HTML `"step"` attribute
`min`											| `integer`											 | Set HTML `"min"` attribute
`max`											| `integer`											 | Set HTML `"max"` attribute

# Global Modifiers

Following DataTable options are supported.

DataTable option      | Accepted values | Description
--------------------- | --------------- | ----------------------------
`altEditor`           | `true \| false` | Enable editor. Default false.
`onAddRow`            | function        | On-add callback function.
`onEditRow`           | function        | On-edit callback function.
`onDeleteRow`         | function        | On-delete callback function.
`closeModalOnSuccess` | `true \| false` | Close modal after successful edit/add/delete. Default true.
`encodeFiles` | `true \| false` | Encode files to Base64 format. Default true.

## Credits

See LICENSE and CHANGELOG for various credits.


CHANGELOG
2021-04 thetechnician94
	*Added more modifiers for number columns
	*Checkboxes will now be checked when the value of a column is 1
	*Dropdowns sometimes got extra spaces around their values, fixed
	*Added the `value` modifier: lets you specify a value for the add dialog
	*Disabled fields are no longer sent to the endpoint. HTTP POST never sends these, maintaining that convention

2020-09 luca-vercelli
    * basic support for Foundation
    * "close modal on success" feature

2020-09 viicslen
    * basic npm support

2020-06 mjbernot
    * support for checkboxes and date input fields
    * IE compatibility support

2020-04 luca-vercelli
    * support for file uploads

2020-02 goalgui10
    * support for textareas

2019-07 luca-vercelli
    * better support for readonly and disabled fields, see issue #35

2019-11 seunets
    * i18n support

2019-05 luca-vercelli
    * better documentation
    * limited support for nested objects, see issue #15

2019-03 luca-vercelli
    * support for datepicker, datetimepicker, select2

2018-09 luca-vercelli
    * some improvements in AJAX callbacks

2018-07 zach-hable
    * Version 2.0
    * Cleaner source code
    * Improved modal layouts
    * Input validation
    * Callbacks for use of AJAX when a row is added/edited/deleted
    * Support for Max length of items
    * Support for Multiple select box
    * Support for using Select2 Dropdowns

2017-07 luca-vercelli
    * Support for object-based data
    * Cleaner source code
    * Refresh button
    * Naive AJAX support

2016-09 KasperOlesen
    * Initial commit
    * Added example and tweaked comments in editor

2016-04 Kingcode
    * Original script found at http://kingkode.com/free-datatables-editor-alternative/
    * It seems that, nowadays, the source code is changed there


example/09_translations/example9.js
$(document).ready(function() {

  var dataSet = [
    [1,"Tiger Nixon", "System Architect", "Edinburgh", "5421", "2011/04/25", "$320,800"],
    [2,"Garrett Winters", "Accountant", "Tokyo", "8422", "2011/07/25", "$170,750"],
    [3,"Ashton Cox", "Junior Technical Author", "San Francisco", "1562", "2009/01/12", "$86,000"],
    [4,"Cedric Kelly", "Senior Javascript Developer", "Edinburgh", "6224", "2012/03/29", "$433,060"],
    [5,"Airi Satou", "Accountant", "Tokyo", "5407", "2008/11/28", "$162,700"],
    [6,"Brielle Williamson", "Integration Specialist", "New York", "4804", "2012/12/02", "$372,000"],
    [7,"Herrod Chandler", "Sales Assistant", "San Francisco", "9608", "2012/08/06", "$137,500"],
    [8,"Rhona Davidson", "Integration Specialist", "Tokyo", "6200", "2010/10/14", "$327,900"],
    [9,"Colleen Hurst", "Javascript Developer", "San Francisco", "2360", "2009/09/15", "$205,500"],
    [10,"Sonya Frost", "Software Engineer", "Edinburgh", "1667", "2008/12/13", "$103,600"],
    [11,"Jena Gaines", "Office Manager", "London", "3814", "2008/12/19", "$90,560"],
    [12,"Quinn Flynn", "Support Lead", "Edinburgh", "9497", "2013/03/03", "$342,000"],
    [13,"Charde Marshall", "Regional Director", "San Francisco", "6741", "2008/10/16", "$470,600"],
    [14,"Haley Kennedy", "Senior Marketing Designer", "London", "3597", "2012/12/18", "$313,500"],
    [15,"Tatyana Fitzpatrick", "Regional Director", "London", "1965", "2010/03/17", "$385,750"],
    [16,"Michael Silva", "Marketing Designer", "London", "1581", "2012/11/27", "$198,500"],
    [17,"Paul Byrd", "Chief Financial Officer (CFO)", "New York", "3059", "2010/06/09", "$725,000"],
    [18,"Gloria Little", "Systems Administrator", "New York", "1721", "2009/04/10", "$237,500"],
    [19,"Bradley Greer", "Software Engineer", "London", "2558", "2012/10/13", "$132,000"],
    [20,"Dai Rios", "Personnel Lead", "Edinburgh", "2290", "2012/09/26", "$217,500"],
    [21,"Jenette Caldwell", "Development Lead", "New York", "1937", "2011/09/03", "$345,000"],
    [22,"Yuri Berry", "Chief Marketing Officer (CMO)", "New York", "6154", "2009/06/25", "$675,000"],
    [23,"Caesar Vance", "Pre-Sales Support", "New York", "8330", "2011/12/12", "$106,450"],
    [24,"Doris Wilder", "Sales Assistant", "Sidney", "3023", "2010/09/20", "$85,600"],
    [25,"Angelica Ramos", "Chief Executive Officer (CEO)", "London", "5797", "2009/10/09", "$1,200,000"],
    [26,"Gavin Joyce", "Developer", "Edinburgh", "8822", "2010/12/22", "$92,575"],
    [27,"Jennifer Chang", "Regional Director", "Singapore", "9239", "2010/11/14", "$357,650"],
    [28,"Brenden Wagner", "Software Engineer", "San Francisco", "1314", "2011/06/07", "$206,850"],
    [29,"Fiona Green", "Chief Operating Officer (COO)", "San Francisco", "2947", "2010/03/11", "$850,000"],
    [30,"Shou Itou", "Regional Marketing", "Tokyo", "8899", "2011/08/14", "$163,000"],
    [31,"Michelle House", "Integration Specialist", "Sidney", "2769", "2011/06/02", "$95,400"],
    [32,"Suki Burks", "Developer", "London", "6832", "2009/10/22", "$114,500"],
    [33,"Prescott Bartlett", "Technical Author", "London", "3606", "2011/05/07", "$145,000"],
    [34,"Gavin Cortez", "Team Leader", "San Francisco", "2860", "2008/10/26", "$235,500"],
    [35,"Martena Mccray", "Post-Sales support", "Edinburgh", "8240", "2011/03/09", "$324,050"],
    [36,"Unity Butler", "Marketing Designer", "San Francisco", "5384", "2009/12/09", "$85,675"]
  ];

  var columnDefs = [{
    title: "Id",
    type: "readonly"
  }, {
    title: "Name",
    type: "text"
  }, {
    title: "Position",
    type: "text"
  }, {
    title: "Office"
    //no type = text
  }, {
    title: "Extn.",
    type: "text"
  }, {
    title: "Start date",
    type: "readonly"
  }, {
    title: "Salary",
    type: "text"
  }];

  var myTable;

  myTable = $('#example').DataTable({
    "sPaginationType": "full_numbers",
    data: dataSet,
    columns: columnDefs,
    dom: 'Bfrtip',        // Needs button container
    select: 'single',
    responsive: true,
    altEditor: true,     // Enable altEditor
    buttons: [{
        text: 'Aggiungi',
        name: 'add'        // do not change name
        },
        {
        extend: 'selected', // Bind to Selected row
        text: 'Modifica',
        name: 'edit'        // do not change name
        },
        {
        extend: 'selected', // Bind to Selected row
        text: 'Elimina',
        name: 'delete'      // do not change name
        }],
    language: {
        
        // see https://datatables.net/plug-ins/i18n/Italian
        "sEmptyTable":     "Nessun dato presente nella tabella",
        "sInfo":           "Vista da _START_ a _END_ di _TOTAL_ elementi",
        "sInfoEmpty":      "Vista da 0 a 0 di 0 elementi",
        "sInfoFiltered":   "(filtrati da _MAX_ elementi totali)",
        "sInfoPostFix":    "",
        "sInfoThousands":  ".",
        "sLengthMenu":     "Visualizza _MENU_ elementi",
        "sLoadingRecords": "Caricamento...",
        "sProcessing":     "Elaborazione...",
        "sSearch":         "Cerca:",
        "sZeroRecords":    "La ricerca non ha portato alcun risultato.",
        "oPaginate": {
            "sFirst":      "Inizio",
            "sPrevious":   "Precedente",
            "sNext":       "Successivo",
            "sLast":       "Fine"
        },
        "oAria": {
            "sSortAscending":  ": attiva per ordinare la colonna in ordine crescente",
            "sSortDescending": ": attiva per ordinare la colonna in ordine decrescente"
        },
        
        // see /translations
        "altEditor" : {
            "modalClose" : "Chiudi",
            "edit" : {
                "title" : "Modifica",
                "button" : "Modifica"
            },
            "add" : {
                "title" : "Nuovo",
                "button" : "Crea"
            },
            "delete" : {
                "title" : "Elimina",
                "button" : "Elimina"
            },
            "success" : "Fatto.",
            "error" : {
                "message" : "Si è verificato un errore non previsto.",
                "label" : "Errore",
                "responseCode" : "Response code:",
                "required" : "Valore obbligatorio",
                "unique" : "Valore duplicato"
            }
        }
    }
  });


});

